### ISSUE TITLE
Possible improvement:  this check depends on scrapy reponse properly converted to XMLReponse intance 

### ISSUE THREAD

**Issue by DharmeshPandav** [scrapy/scrapy#6362]
This check here depends on other modules and relies on if scrapy response had been converted to an instance of "XmlResponse"

https://github.com/scrapy/scrapy/blob/b88f22c6c5de4ca8828b2abe860516c2468c4fe3/scrapy/spiders/sitemap.py#L109

The assumption here that the response will be typecasted to XmlResponse, might not be true in all cases and Sitemap Spider will stop working as expected. 

For Example: Zyte-API responses are always of a type `ZyteAPITextResponse` hence Sitemap spider will not work with zyte-api out of the box.

have checked and we have a code [here](https://github.com/scrapy/scrapy/blob/b88f22c6c5de4ca8828b2abe860516c2468c4fe3/scrapy/responsetypes.py#L19) to assess the response types based on headers, body and url which should be used instead

suggestion here is that , we can either put this in addition to the existing check at the end or replace the first check 

```
if responsetypes.from_args(headers=response.headers, url=response.url, body=response.body) == XmlResponse:
 return response.body
```

**Comment by Gallaecio** [comment]
I’m not sure if this is something that Scrapy should handle, maybe scrapy-zyte-api should have alternative subclasses for different Response subclasses that also inherit from them.

**Comment by DharmeshPandav** [comment]
Agreed that core code should be 3rd party library agnostic. This should be handled in downstream code to make it compatible with core.

The suggestion is to de-couple the check and make it less cohesive and explicit when checking whether the response is a sitemap or not. and can be done by re-using the existing module.

**Comment by Gallaecio** [comment]
The main issue I see with the proposal is that the potentially non-cheap MIME type check from responsetypes is done more than once, once in the Scrapy core and once here.

That said, since it is only done for sitemap responses, which are often only a small fraction of a crawl, maybe that’s OK performance-wise. Although it still feels slightly ugly.
